include/deal.II/grid/manifold_lib.h

```diff
@@ -503,6 +504,282 @@ class TorusManifold : public ChartManifold<dim,3,3>
+   */
+  virtual
+  void
+  add_new_points (const std::vector<Point<spacedim> > &surrounding_points,
+                  const Table<2,double>               &weights,
+                  std::vector<Point<spacedim> >       &new_points) const;
+
+private:
+  /**
+   * Internal function to identify the most suitable cells (=charts) where the
+   * given surrounding points are located. We use a cheap algorithm to identify
+   * the cells and rank the cells by probability before we actually do the
+   * search inside the relevant cells. We expect at most 10 cells, so get an
+   * array with 10 entries of a the indices <tt>cell->index()</tt>.
+   */
+  std::array<unsigned int, 10>
+  get_possible_cells_around_points(const std::vector<Point<spacedim> > &surrounding_points) const;
+
+  /**
+   * Finalizes the identification of the correct chart and returns the location
+   * of the surrounding points on the chart. This method internally calls
+   * @p get_possible_cells_around_points
+   */
+  std::pair<typename Triangulation<dim,spacedim>::cell_iterator,
+      std::vector<Point<dim> > >
+      compute_chart_points(const std::vector<Point<spacedim> > &surrounding_points) const;
+
+  /**
+   * Pull back operation into the unit coordinates on the given coarse cell.
+   *
+   * @note This internal function is currently not compatible with the
+   * ChartManifold::pull_back() function because the given class represents an
+   * atlas of charts, not a single chart. Thus, the pull_back() operation is
+   * only valid with the additional information of the chart, given by a cell
+   * on the coarse grid. An alternative implementation could shift the index
+   * depending on the coarse cell for a 1-to-1 relation between the chart space
+   * and the image space.
+   */
+  Point<dim>
+  pull_back(const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+            const Point<spacedim> &p) const;
+
+  /**
+   * Push forward operation.
+   *
+   * @note This internal function is currently not compatible with the
+   * ChartManifold::pull_back() function because the given class represents an
+   * atlas of charts, not a single chart. Thus, the pull_back() operation is
+   * only valid with the additional information of the chart, given by a cell
+   * on the coarse grid. An alternative implementation could shift the index
+   * depending on the coarse cell for a 1-to-1 relation between the chart space
+   * and the image space.
+   */
+  Point<spacedim>
+  push_forward(const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+               const Point<dim> &chart_point) const;
+
+  /**
+   * Gradient of the push_forward method.
+   */
+  DerivativeForm<1,dim,spacedim>
+  push_forward_gradient(const typename Triangulation<dim,spacedim>::cell_iterator &cell,
+                        const Point<dim> &chart_point) const;
+
+  /**
+   * The underlying triangulation.
+   */
+  const Triangulation<dim,spacedim> *triangulation;
```

Comment:
or maybe not... This would give us troubles for the fact that `Manifolds` are referenced by `Triangulation`, and closing the loop will always throw an error, no matter who we destroy first, unless we explicitly clear the pointer from the `Triangulation`... just ignore my comment above...